waiter/src/waiter/metrics_consumer.clj

```diff
@@ -0,0 +1,193 @@
+  /token-stats?watch=true endpoint. Each json blob has the structure:
+  {:object [
+    {:token \"token\"
+     :lastRequestTime iso-time-string}
+    ...]
+   :type \"initial\" or \"update\"}
+
+  A map is returned with keys:
+  :exit-chan is a channel that when a message is pushed to, terminates the watch immediately and go-chan
+  :go-chan channel closed when watch is terminated (used primarily for cleaning up tests)
+  :query-state-fn function called with include flags for introspecting the state of the watch (e.g. last-event-time)"
+  [http-client http-streaming-request-async-fn router-id cur-cid {:keys [cluster url] :as metrics-service-conf}
+   event-chan clock retry-delay-ms]
+  (cid/with-correlation-id
+    (str cur-cid ".metrics-watch." cluster)
+    (let [_ (log/info "starting metrics-service-watch" {:metrics-service-conf metrics-service-conf})
+          exit-chan (async/promise-chan)
+          exit-chan-mult (async/mult exit-chan)
+          state-atom (atom {:last-event-time nil
+                            :last-watch-time nil})
+          metrics-endpoint (str url "/token-stats")
+          query-string (str "watch=true&watcher=" router-id)
+          query-state-fn
+          (fn query-metrics-watch-fn
+            [_]
+            (let [{:keys [last-event-time last-watch-time]} @state-atom]
+              {:endpoint metrics-endpoint
+               :last-event-time last-event-time
+               :last-watch-time last-watch-time
+               :supported-include-params []}))
+          go-chan
+          (async/go-loop []
+            (try
+              (log/info "starting watch on metrics service:" {:metrics-endpoint metrics-endpoint})
+              (let [abort-ch (async/promise-chan)
+                    {:keys [body-chan error-chan] :as res}
+                    (async/<! (http-streaming-request-async-fn http-client metrics-endpoint
+                                                               :abort-ch abort-ch
+                                                               :query-string query-string))]
+                ; when async chan is triggered, we want to also trigger the abort channel on the watch connection
+                (async/tap exit-chan-mult abort-ch)
+                (when (instance? Throwable res)
+                  (throw res))
+                (swap! state-atom assoc :last-watch-time (clock))
+                (doseq [{:strs [object type]} (utils/chan-to-json-seq!! body-chan)]
+                  (meters/mark! (metrics/waiter-meter "core" "metrics-consumer" cluster "event-rate"))
+                  (if (contains? #{"initial" "update"} type)
+                    (do
+                      (log/info "received event payload" {:count (count object) :type type})
+                      (swap! state-atom assoc :last-event-time (clock))
+                      (doseq [event object]
+                        (async/put! event-chan event)))
```

Comment:
Can we avoid a loop to put each individual event into the channel? We should be able to send the collections of events and get it processed donwstream.